### PR TITLE
Document fetchTree

### PR DESCRIPTION
**Pull request by wentasah**
# Motivation

WiP commit originally included in #8491.

# Context

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Review by thufschmitt — COMMENTED** [review]
Big :+1: on the idea, though I'll defer to the doc team for reviewing the content

**Review by fricklerhandwerk — APPROVED** [review]
Please reformat to one sentence per line, as this will make changes in the future much less noisy.

File: src/libexpr/primops/fetchTree.cc [hunk 1]
```diff
@@ -193,10 +193,50 @@ static void prim_fetchTree(EvalState & state, const PosIdx pos, Value * * args,
...
 
-// FIXME: document
 static RegisterPrimOp primop_fetchTree({
     .name = "fetchTree",
-    .arity = 1,
+    .args = {"args"},
+    .doc = R"(
+      Fetch a source tree or a plain file using one of the supported
+      backends. *args* can be an attribute set representation of
+      [flake reference](../command-ref/new-cli/nix3-flake.html#flake-references)
```
Comment (fricklerhandwerk) [inline src/libexpr/primops/fetchTree.cc:202]:
```suggestion
      [flake reference](@docroot@/command-ref/new-cli/nix3-flake.md#flake-references)
```

File: src/libexpr/primops/fetchTree.cc [hunk 2]
```diff
@@ -193,10 +193,50 @@ static void prim_fetchTree(EvalState & state, const PosIdx pos, Value * * args,
...
+          ```
+      - Fetch a single file from a URL:
+
+          ```nix
+          builtins.fetchTree "https://example.com/"
+          ```
+
+      > **Note**
+      >
+      > This function requires the [`flakes` experimental feature](@docroot@/contributing/experimental-features.md#xp-feature-flakes) to be enabled and the requested flake reference to be locked in `flake.lock`.
```
Comment (fricklerhandwerk) [inline src/libexpr/primops/fetchTree.cc:238]: @Ericson2314 is the automatic experimental admonition already merged?

**Review by fricklerhandwerk — APPROVED** [review]
This looks good. Not sure if the experimental guard can be already automated. If not, ready to merge from my side.

**Comment by wentasah** [comment]
I'll get to finish this tomorrow. Sorry for delays.

**Comment by wentasah** [comment]
I made the following changes:
- Formatted as one sentence per line.
- Added a sentence about locked input (previous mention of `flake.lock` was incorrect). The question is whether it should be called *locked* (as in the [source code](https://github.com/wentasah/nix/blob/f11445952fcf0f5fd3f827997591d43a41f2074a/src/libfetchers/fetchers.hh#L67-L71)) or *immutable* as in [a recent @edolstra PR](https://github.com/NixOS/nix/pull/8477). I'd prefer *immutable*, in which case, we should also change [this error message](https://github.com/wentasah/nix/blob/f11445952fcf0f5fd3f827997591d43a41f2074a/src/libexpr/primops/fetchTree.cc#L186).
- Changed the first example to use "locked" input.

File: src/libexpr/primops/fetchTree.cc (hunk 2 above)
Comment (Ericson2314) [reply src/libexpr/primops/fetchTree.cc]: It is now!

Comment (Ericson2314) [reply src/libexpr/primops/fetchTree.cc]: Let's fix this later.

**Pull request merged**